src/libraries/System.Runtime/tests/System/Reflection/AssemblyTests.cs

```diff
@@ -0,0 +1,1563 @@
+            yield return new object[] { "Type", true, false, typeof(List<int>) };
+            yield return new object[] { "Type", false, true, typeof(List<int>) };
+            yield return new object[] { "Type", false, false, typeof(List<int>) };
+        }
+
+        [Theory]
+        [MemberData(nameof(GetType_String_Bool_BoolCustom_TestData))]
+        public void GetType_InvokeStringBoolBoolCustom_ReturnsExpected(string name, bool throwOnError, bool ignoreCase, Type result)
+        {
+            var assembly = new CustomAssembly
+            {
+                GetTypeAction = (nameParam, throwOnErrorParam, ignoreCaseParam) =>
+                {
+                    Assert.Equal(name, nameParam);
+                    Assert.Equal(throwOnError, throwOnErrorParam);
+                    Assert.Equal(ignoreCase, ignoreCaseParam);
+                    return result;
+                }
+            };
+            Assert.Same(result, assembly.GetType(name, throwOnError, ignoreCase));
+        }
+
+        [Fact]
+        public void GetTypes_Invoke_ThrowsNotImplementedException()
+        {
+            var assembly = new SubAssembly();
+            Assert.Throws<NotImplementedException>(() => assembly.GetTypes());
+        }
+
+        public static IEnumerable<object[]> GetTypes_Custom_TestData()
+        {
+            yield return new object[] { new Module[] { typeof(int).Module }, typeof(int).Module.GetTypes() };
+
+            var module1 = new CustomModule
+            {
+                GetTypesAction = () => new Type[0]
+            };
+            var module2 = new CustomModule
+            {
+                GetTypesAction = () => new Type[] { typeof(int), typeof(string) }
+            };
+            var module3 = new CustomModule
+            {
+                GetTypesAction = () => new Type[] { null, typeof(bool) }
+            };
+            yield return new object[] { new Module[] { module2 }, new Type[] { typeof(int), typeof(string) } };
+            yield return new object[] { new Module[] { new CustomModule { GetTypesAction = () => null } }, null };
+            yield return new object[] { new Module[] { module1, module2, module3 }, new Type[] { typeof(int), typeof(string), null, typeof(bool) } };
+        }
+
+        [Theory]
+        [MemberData(nameof(GetTypes_Custom_TestData))]
+        public void GetTypes_InvokeCustom_ReturnsExpected(Module[] modules, Type[] expected)
+        {
+            var assembly = new CustomAssembly
+            {
+                GetModulesAction = getResourceModules => modules
+            };
+            Assert.Equal(expected, assembly.GetTypes());
+        }
+
+        [Fact]
+        public void GetTypes_NullGetModules_ThrowsNullReferenceException()
+        {
+            var assembly = new CustomAssembly
+            {
+                GetModulesAction = (getResourceModules) => null
+            };
+            Assert.Throws<NullReferenceException>(() => assembly.GetTypes());
```

Comment:
its good to know because it could be a bug. Any changes in the future could regress something

Reply:
I don't think handling NRE differently could be a bug but @steveharter's might have a different opinion

Reply:
@hughbe the tests should be documenting the expected behavior. NRE is never an expected exception. I'd recommend to create an issue with all of the cases where you found NRE and then we should fix it separately. You can add all tests into the issue so that it's easier to add it back.

Reply:
I think it could be a bug if it is reproducable with any realistic scenarios like the parameter was not set properly etc but in this case, it unconditionally returns null which is not an expected scenario

Reply:
How does `GetModulesAction` cause the NRE to be thrown? I assume the base class (`System.Reflection.Assembly`) is throwing it?

Should the test be setting `GetTypesAction => null` instead of `GetModulesAction`?

Reply:
If you look here: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Reflection/Assembly.cs#L45-L72

`GetTypes` calls `GetModules`. If `GetModules` returns `null`, then an NRE is thrown

```
        public virtual Type[] GetTypes()
        {
            Module[] m = GetModules(false);
            if (m.Length == 1)
            {
                return m[0].GetTypes();
            }

            int finalLength = 0;
            Type[][] moduleTypes = new Type[m.Length][];

            for (int i = 0; i < moduleTypes.Length; i++)
            {
                moduleTypes[i] = m[i].GetTypes();
                finalLength += moduleTypes[i].Length;
            }

            int current = 0;
            Type[] ret = new Type[finalLength];
            for (int i = 0; i < moduleTypes.Length; i++)
            {
                int length = moduleTypes[i].Length;
                Array.Copy(moduleTypes[i], 0, ret, current, length);
                current += length;
            }

            return ret;
        }

```

Reply:
I'm ambivalent about fixing this - these would be rare cases that are not really expected or in the contract of the API.

I do like the idea of keeping these tests though as it explicitly calls out behaviour in edge cases. If we chose not to fix an NRE, for example, then this decision is explicitly documented through testing.